examples/src/main/java/com/squareup/moshi/recipes/RuntimeJsonAdapterFactory.java

```diff
@@ -0,0 +1,129 @@
+    }
+    this.baseType = baseType;
+    this.labelKey = labelKey;
+  }
+
+  public RuntimeJsonAdapterFactory registerSubtype(Class<?> subtype, String label) {
+    if (subtype == null) {
+      throw new NullPointerException("subtype == null");
+    }
+    if (label == null) {
+      throw new NullPointerException("label == null");
+    }
+    if (!baseType.isAssignableFrom(subtype)) {
+      throw new IllegalArgumentException(subtype + " must be a " + baseType);
+    }
+    subtypeToLabel.put(subtype, label);
+    return this;
+  }
+
+  @Override
+  public JsonAdapter<?> create(Type type, Set<? extends Annotation> annotations, Moshi moshi) {
+    if (!annotations.isEmpty() || Types.getRawType(type) != baseType) {
+      return null;
+    }
+    Map<Class<?>, String> subtypeToLabel = new LinkedHashMap<>(this.subtypeToLabel);
+    int size = subtypeToLabel.size();
+    Map<String, JsonAdapter<?>> labelToDelegate = new LinkedHashMap<>(size);
+    Map<Class<?>, JsonAdapter<?>> subtypeToDelegate = new LinkedHashMap<>(size);
+    for (Map.Entry<Class<?>, String> entry : subtypeToLabel.entrySet()) {
+      Class<?> key = entry.getKey();
+      String value = entry.getValue();
+      JsonAdapter<?> delegate = moshi.adapter(key, annotations);
+      labelToDelegate.put(value, delegate);
+      subtypeToDelegate.put(key, delegate);
+    }
+    JsonAdapter<Map<String, Object>> toJsonDelegate =
+        moshi.adapter(Types.newParameterizedType(Map.class, String.class, Object.class));
+    return new RuntimeJsonAdapter(labelKey, labelToDelegate, subtypeToDelegate, subtypeToLabel,
+        toJsonDelegate);
+  }
+
+  private static final class RuntimeJsonAdapter extends JsonAdapter<Object> {
+    private final String labelKey;
+    private final Map<String, JsonAdapter<?>> labelToDelegate;
+    private final Map<Class<?>, JsonAdapter<?>> subtypeToDelegate;
+    private final Map<Class<?>, String> subtypeToLabel;
+    private final JsonAdapter<Map<String, Object>> toJsonDelegate;
+
+    RuntimeJsonAdapter(String labelKey, Map<String, JsonAdapter<?>> labelToDelegate,
+        Map<Class<?>, JsonAdapter<?>> subtypeToDelegate, Map<Class<?>, String> subtypeToLabel,
+        JsonAdapter<Map<String, Object>> toJsonDelegate) {
+      this.labelKey = labelKey;
+      this.labelToDelegate = labelToDelegate;
+      this.subtypeToDelegate = subtypeToDelegate;
+      this.subtypeToLabel = subtypeToLabel;
+      this.toJsonDelegate = toJsonDelegate;
+    }
+
+    @Override public Object fromJson(JsonReader reader) throws IOException {
+      Object raw = reader.readJsonValue();
+      if (!(raw instanceof Map)) {
+        throw new JsonDataException(
+            "Value must be a JSON object but had a value of " + raw + " of type " + raw.getClass());
+      }
+      @SuppressWarnings("unchecked") // This is a JSON object.
+          Map<String, Object> value = (Map<String, Object>) raw;
+      Object label = value.remove(labelKey);
```

Comment:
Yep, this is only to emulate Gson's RuntimeTypeAdapterFactory: https://github.com/google/gson/blob/master/extras/src/main/java/com/google/gson/typeadapters/RuntimeTypeAdapterFactory.java#L202

Copy, paste, modify to your needs.